### PR TITLE
feat: improve track grade parsing

### LINKED ISSUE #1774: Improve track grade parsing
## Problem description

When parsing combined OSM tag values of [tracktype](https://wiki.openstreetmap.org/wiki/Key:tracktype) currently only values listed with `;` are taken into account but not ranges listed with `-`. However, according to [taginfo](https://taginfo.openstreetmap.org/keys/tracktype#values) there is an order of magnitude more occurrences of the latter ones than the former ones.

``` r
# query data from taginfo
url <- "https://taginfo.openstreetmap.org/api/4/key/values?key=tracktype&filter=ways&sortname=count&sortorder=desc&qtype=value"
res <- jsonlite::fromJSON(url)$data

# number of combined range-type values
with(res, count[grepl("-", value) & grepl("grade", value)] |> sum())
#> [1] 3493

# number of combined semicolon-separated values
with(res, count[grepl(";", value) & grepl("grade", value)] |> sum())
#> [1] 296
```
<sup>Created on 2024-04-05 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

### Other common easy to address tagging mistakes

``` r
# query data from taginfo
url <- "https://taginfo.openstreetmap.org/api/4/key/values?key=tracktype&filter=ways&sortname=count&sortorder=desc&qtype=value"
res <- jsonlite::fromJSON(url)$data

# single-number values such as `tracktype=5`
with(res, count[grepl("^[1-5]$", value)]  |> sum())
#> [1] 125

# higher grades 
res[with(res, which(grepl("^grade[6-9]+$", value))),1:2]
#>     value count
#> 8  grade6   200
#> 10 grade7    71
#> 38 grade8     4
```

<sup>Created on 2024-04-05 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

### Current code for parsing `tracktype` values

https://github.com/GIScience/openrouteservice/blob/78ec9adf6bc9214417fa3eca7d35c76ea9a1d4a8/ors-engine/src/main/java/org/heigit/ors/routing/graphhopper/extensions/flagencoders/VehicleFlagEncoder.java#L420-L457

## Proposed solution

Include parsing of grade range values provided with `-`.

```java
    protected int getTrackGradeLevel(String grade) {
        if (grade == null) 
            return 0; 
        switch (grade) {
            // official values
            case "grade1":
                return 1;
            case "grade2":
                return 2;
            case "grade3":
                return 3;
            case "grade4":
                return 4;
            case "grade5":
                return 5;
            // attempt to extract meaningful values from common tagging mistakes
            case "grade6":
            case "grade7":
            case "grade8":
                return 6;
            default:
                // call specialized parsing function:
                //    split on ";" or "-"
                //    remove "grade" and trim whitespaces
                //    parse as integer and take maximum over these values and 6
                //
                //    bonus: parse single-number values such as `tracktype=5`
        }
        return 0;
    }
```

### PR DESCRIPTION
**Pull request by hwiks**
### Pull Request Checklist

- [X] 1. I have [**rebased**][rebase] the latest version of the main branch into my feature branch and all conflicts
         have been resolved.
- [] 2. I have added information about the change/addition to functionality to the CHANGELOG.md file under the
         [Unreleased] heading.
- [not applicable]  3. I have documented my code using JDocs tags.
- [X] 4. I have removed unnecessary commented out code, imports and System.out.println statements.
- [X] 5. I have written JUnit tests for any new methods/classes and ensured that they pass.
- [not applicable] 6. I have created API tests for any new functionality exposed to the API.
- [not applicable] 7. If changes/additions are made to the ors-config.json file, I have added these to the [ors config documentation][config]
         along with a short description of what it is for, and documented this in the Pull Request (below).
- [not applicable] 8. I have built graphs with my code of the Heidelberg.osm.gz file and run the api-tests with all test passing
- [X] 9. I have referenced the Issue Number in the Pull Request (if the changes were from an issue).
- [not applicable] 10. For new features or changes involving building of graphs, I have tested on a larger dataset
          (at least Germany), and the graphs build without problems (i.e. no out-of-memory errors).
- [not applicable] 11. For new features or changes involving the graphbuilding process (i.e. changing encoders, updating the
          importer etc.), I have generated longer distance routes for the affected profiles with different options
          (avoid features, max weight etc.) and compared these with the routes of the same parameters and start/end
          points generated from the current live ORS.
          If there are differences then the reasoning for these **MUST** be documented in the pull request.
- [X] 12. I have written in the Pull Request information about the changes made including their intended usage
          and why the change was needed.
- [not applicable] 13. For changes touching the API documentation, I have tested that the API playground [renders correctly][api].

Fixes #1774.

### Information about the changes
- Parsing of lists of track grades now work when lists are separated by '-'
- Parsing of grades between 6 and 8 are now all parsed as 6 to extract some meaningful analysis
- Parsing of lists of track grades now follow rules for single grade parsing (6,7,8 return 6, higher than 8 or not numbers return 1)
- tests are now implemented for getTrackGradeLevel, documenting the expected behaviours
- Reason for change: lists of grades using - were not parsed, and parsing of lists did not follow single parsing rules. Also, no tests were implemented yet for the function in the abstract class VehicleFlagEncoder 

NOTE: This is my first PR in this project, please review with extra care and especially please double check that my tests make the correct assumptions about desired behavior!! 

### Examples and reasons for differences between live ORS routes, and those generated from this pull request
- not applicable

### Required changes to ors config (if applicable)
- not applicable

[config]: https://GIScience.github.io/openrouteservice/run-instance/configuration/
[api]: https://gitlab.gistools.geog.uni-heidelberg.de/giscience/openrouteservice-infrastructure/ors-docs-api#test-new-ors-documentation
[rebase]: https://github.com/GIScience/openrouteservice/blob/main/CONTRIBUTE.md#pull-request-guidelines

### CODE REVIEW

**Comment by koebi** [comment]
Hey,

thanks a lot for your contribution, and especially thanks for creating tests!

I have not looked at the code explicitly, but rather at the `tracktype`-key itself, and I am curious:

When looking at [the OSM wiki description for the `tracktype` key](https://wiki.openstreetmap.org/wiki/Key:tracktype), this only lists `grade1` through `grade5`, and according to [the taginfo page for `tracktype`](https://taginfo.openstreetmap.org/keys/tracktype#values), these cover 99.95% of values.

I would love to see this aligned in the OSM a bit better, rather than complicating the tag parsing of the openrouteservice, but there might be good reason for the openrouteservice to handle this.

Could you provide an example on where the current routing fails, i.e. what caused you to add this PR?

Last, as a minor nitpick: 
We do use conventional commits - see https://www.conventionalcommits.org/ - so it would be great if you had the time to re-name your commits accordingly :) If not, we will take care of it, no worries.

Best regards :)

**Comment by hwiks** [comment]
Thank you for your comment! 

For your question about the grade parsing: I did none of the thinking. I executed what @aoles describes in their ticket here: https://github.com/GIScience/openrouteservice/issues/1774

It looks like the problem for them is that many mistakes could be bundled up and analyzed as a non-existent tag '6', and they provide a count. 

Regardless, at the moment the lists are handled in such a way that the returned tag can be any number as long as it's the higher number in the list, and that is not, to my understanding, meaningful behavior. 

I'm very happy to change the logic to whatever your group agrees on! Let me know if there is an agreement :) 

As for the commit message, no problem! I'll squash all commits into one with the right format once the logic is defined :)

Thank you so much for taking the time!

**Comment by aoles** [comment]
Thanks a lot @hwiks for your contribution and patience! ❤️ 

We have had a discussion it the team and came up with a somewhat standardised and simplified approach. The general idea is to stick with the values returned by `getTrackGradeLevel()` to ones corresponding to the official set of values ranging from `grade1` to `grade5`. The will find the updated proposal in the [comment on the issue](https://github.com/GIScience/openrouteservice/issues/1774#issuecomment-3991367445).

Looking very much forward to reviewing your modified code 🤓 

Cheers!
Andrzej

**Comment by hwiks** [comment]
Hello! Thank you for the clear added comments: I added the changes and updated the tests, as well as committed in the conventional commit format. Hopefully all is correct, the tests should also clarify how I expect the code to behave in some edge cases, such as: a list of grades where one of them is unparsable should return 0. If you think those edge cases should be handled differently, let me know and I'll update it 👍

File: ors-engine/src/test/java/org/heigit/ors/routing/graphhopper/extensions/flagencoders/VehicleFlagEncoderTest.java [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+    void shouldReturn5WhenGrade7() {
+        int result = service.getTrackGradeLevel("grade7");
+        assertEquals(5, result);
+    }
+
+    @Test
+    void shouldReturn5WhenGrade8() {
+        int result = service.getTrackGradeLevel("grade8");
+        assertEquals(5, result);
+    }
```
Comment (aoles) [inline ors-engine/src/test/java/org/heigit/ors/routing/graphhopper/extensions/flagencoders/VehicleFlagEncoderTest.java]: Instead of having separate test for grades higher than 5 you might consider implementing a parametrized test `shouldReturn5WhenGradeHigherThan5`.

File: ors-engine/src/test/java/org/heigit/ors/routing/graphhopper/extensions/flagencoders/VehicleFlagEncoderTest.java [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+    @Test
+    void shouldReturn4WhenGrade4() {
+        int result = service.getTrackGradeLevel("grade4");
+        assertEquals(4, result);
+    }
+    @Test
+    void shouldReturn5WhenGrade5() {
+        int result = service.getTrackGradeLevel("grade5");
+        assertEquals(5, result);
+    }
```
Comment (aoles) [inline ors-engine/src/test/java/org/heigit/ors/routing/graphhopper/extensions/flagencoders/VehicleFlagEncoderTest.java]: Please use consistent code formatting: you might want to separate individual test methods by a single blank line.

File: ors-engine/src/test/java/org/heigit/ors/routing/graphhopper/extensions/flagencoders/VehicleFlagEncoderTest.java [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+    void shouldReturn0WhenGradeAllCaps() {
+        int result = service.getTrackGradeLevel("GRADE");
+        assertEquals(0, result);
+    }
+
+    @Test
+    void shouldReturn1WhenGrade() {
+        int result = service.getTrackGradeLevel("grade");
+        assertEquals(1, result);
+    }
```
Comment (aoles) [inline ors-engine/src/test/java/org/heigit/ors/routing/graphhopper/extensions/flagencoders/VehicleFlagEncoderTest.java]: Actually, the value `"grade"` does not provide a valid grade specification. Therefore, the expected result should be `0` rather than `1`.

File: ors-engine/src/test/java/org/heigit/ors/routing/graphhopper/extensions/flagencoders/VehicleFlagEncoderTest.java [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+    void shouldReturn5WhenMultipleGradesAndValues678() {
+        int result = service.getTrackGradeLevel("grade1-grade8-grade7");
+        assertEquals(5, result);
+    }
+
+    @Test
+    void shouldReturn5WhenMultipleGradesAndValuesHigherThan8() {
+        int result = service.getTrackGradeLevel("grade1-grade8-grade77");
+        assertEquals(5, result);
+    }
```
Comment (aoles) [inline ors-engine/src/test/java/org/heigit/ors/routing/graphhopper/extensions/flagencoders/VehicleFlagEncoderTest.java:88]: None of the test values corresponds to actual values encountered in OSM, see [taginfo](https://taginfo.openstreetmap.org/keys/tracktype#values). Grade range values have the form `"grade<FROM>-<TO>"` such as `"grade3-5"`. Note that the range values involve only two values `FROM` and `TO`. Please adapt the tests accordingly.

File: ors-engine/src/test/java/org/heigit/ors/routing/graphhopper/extensions/flagencoders/VehicleFlagEncoderTest.java [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+
+    @Test
+    void shouldReturnZeroWhenGradeIsUnknown() {
+        int result = service.getTrackGradeLevel("unknown");
+        assertEquals(0, result);
+    }
+
+    @Test
+    void shouldReturnMaxGradeWhenMultipleGradesPresentSemicolon() {
+        int result = service.getTrackGradeLevel("grade1; grade3; grade2");
```
Comment (aoles) [inline ors-engine/src/test/java/org/heigit/ors/routing/graphhopper/extensions/flagencoders/VehicleFlagEncoderTest.java]: As a test value you might consider using one of the most frequent actual values encountered in OSM e.g. [`"grade3;grade2"`](https://taginfo.openstreetmap.org/tags/tracktype=grade3%3Bgrade2).

File: ors-engine/src/main/java/org/heigit/ors/routing/graphhopper/extensions/flagencoders/VehicleFlagEncoder.java [hunk 6]
```diff
@@ -452,11 +436,50 @@ protected int getTrackGradeLevel(String grade) {
...
+            if (parsed > maxGrade) {
+                maxGrade = parsed;
+            }
+        }
+
+        if (maxGrade <6)
+            return maxGrade;
+        else {
+            return Math.min(5, maxGrade);
+        }
```
Comment (aoles) [inline ors-engine/src/main/java/org/heigit/ors/routing/graphhopper/extensions/flagencoders/VehicleFlagEncoder.java]: The whole `if ... else` block is obsolete and could be reduced to just `return Math.min(5, maxGrade)`.

File: ors-engine/src/main/java/org/heigit/ors/routing/graphhopper/extensions/flagencoders/VehicleFlagEncoder.java [hunk 7]
```diff
@@ -452,11 +436,50 @@ protected int getTrackGradeLevel(String grade) {
...
+
+        // Split by ";" or "-"
+        String[] values = grade.split("[;-]");
+    
+        int maxGrade = 0;
+    
+        for (String v : values) {
+            int parsed = 0;
+            try {
+                parsed = parseSingleGrade(v);
```
Comment (aoles) [inline ors-engine/src/main/java/org/heigit/ors/routing/graphhopper/extensions/flagencoders/VehicleFlagEncoder.java]: Not sure it's worth introducing a separate method unless there was some reason behind it. You might consider inlining `parseSingleGrade` as in the original code:
```java
parsed = Integer.parseInt(v.replace("grade", "").trim());`.
```

File: ors-engine/src/main/java/org/heigit/ors/routing/graphhopper/extensions/flagencoders/VehicleFlagEncoder.java [hunk 8]
```diff
@@ -452,11 +436,50 @@ protected int getTrackGradeLevel(String grade) {
...
-        return 10;
+    }
+                
+    private int guessTrackGrade(String grade) {
+        switch(grade){
+            case "grade":
+                return 1;
+            default: 
+                return parseMaxTrackGrade(grade);
+        }
```
Comment (aoles) [inline ors-engine/src/main/java/org/heigit/ors/routing/graphhopper/extensions/flagencoders/VehicleFlagEncoder.java]: No need of this switch statement and a separate `parseMaxTrackGrade` method. Please inline the code of `parseMaxTrackGrade` here.

**Review by aoles — CHANGES_REQUESTED** [review]
Hi @hwiks,
thanks a lot for updating your code according to my recommendations ❤️ 

Please find a few additional suggestions for further improvements below.

Cheers!
Andrzej

**Comment by aoles** [comment]
Hi @hwiks ,

Thanks again for your contribution ❤️ Just a friendly reminder: please have a look at my recent comments. It would be great if we could merge your PR soon 🚀

Cheers!

**Comment by hwiks** [comment]
> Hi @hwiks ,
> 
> Thanks again for your contribution ❤️ Just a friendly reminder: please have a look at my recent comments. It would be great if we could merge your PR soon 🚀
> 
> Cheers!

Thank you for your comments and patience! I was sick for a while, I will go over everything today 👍

File: ors-engine/src/test/java/org/heigit/ors/routing/graphhopper/extensions/flagencoders/VehicleFlagEncoderTest.java (hunk 1 above)
Comment (hwiks) [reply ors-engine/src/test/java/org/heigit/ors/routing/graphhopper/extensions/flagencoders/VehicleFlagEncoderTest.java]: nifty! thank you, I didn't know about it

File: ors-engine/src/main/java/org/heigit/ors/routing/graphhopper/extensions/flagencoders/VehicleFlagEncoder.java (hunk 6 above)
Comment (hwiks) [reply ors-engine/src/main/java/org/heigit/ors/routing/graphhopper/extensions/flagencoders/VehicleFlagEncoder.java]: great point 🤦

File: ors-engine/src/main/java/org/heigit/ors/routing/graphhopper/extensions/flagencoders/VehicleFlagEncoder.java (hunk 7 above)
Comment (hwiks) [reply ors-engine/src/main/java/org/heigit/ors/routing/graphhopper/extensions/flagencoders/VehicleFlagEncoder.java]: just meant to make the code more readable, but no problem 👍

**Review by aoles — APPROVED** [review]
LGTM 🚀 

Thanks again for your contribution and patience throughout the review process ❤️
